web/src/engine/websites/TappyToon.ts

```diff
@@ -0,0 +1,108 @@
+import { Tags } from '../Tags';
+import icon from './TappyToon.webp';
+import { Chapter, DecoratableMangaScraper, Manga, type MangaPlugin, Page } from '../providers/MangaPlugin';
+import * as Common from './decorators/Common';
+import { FetchJSON, FetchWindowScript } from '../platform/FetchProvider';
+
+type NextData = {
+    props: {
+        initialProps: {
+            pageProps: {
+                comic: APIManga
+            }
+        },
+        initialState: {
+            axios: {
+                headers: {
+                    Authorization: string,
+                    'X-Device-Uuid': string
+                }
+            }
+        }
+    }
+}
+
+type APIManga = {
+    id: number,
+    title: string,
+    isSuperHighQualitySupported: boolean
+}
+
+type APIChapter = {
+    id: number,
+    title: string,
+    isAccessible: boolean,
+    isFree: boolean,
+    isUserUnlocked: boolean,
+    isUserRented: boolean,
+}
+
+type APIPages = {
+    files: {
+        url: string
+    }[]
+}
+
+@Common.ImageAjax(true)
+export default class extends DecoratableMangaScraper {
+    protected language = 'en';
+    private bearer: string = undefined;
+    private deviceid = undefined;
+    private readonly apiUrl = 'https://api-global.tappytoon.com';
+
+    public constructor(id = 'tappytoon', label = 'TappyToon', url = 'https://www.tappytoon.com', tags = [Tags.Media.Manhwa, Tags.Language.English, Tags.Source.Official]) {
+        super(id, label, url, ...tags);
+    }
+
+    public override get Icon() {
+        return icon;
+    }
+
+    public override ValidateMangaURL(url: string): boolean {
+        return new RegExpSafe(`^${this.URI.origin}/${this.language}/book/[^/]+$`).test(url);
+    }
+
+    public override async Initialize(): Promise<void> {
+        const { props: { initialState: { axios: { headers } } } } = await this.GetNextData(new URL(this.URI));
+        this.bearer = headers.Authorization;
+        this.deviceid = headers['X-Device-Uuid'];
+    }
+
+    public override async FetchManga(provider: MangaPlugin, url: string): Promise<Manga> {
+        const { props: { initialProps: { pageProps: { comic } } } } = await this.GetNextData(new URL(url));
+        return new Manga(this, provider, comic.id.toString(), comic.title.replace(/(\[(FR|DE)\])$/, '').trim());
```

Comment:
Everytime a new TappyToon website is added this regex must be changed.
Consider a more generic approach, ideally each inherited website is providing the correct title cleanup.

Or make a single website with support for all languages.

Reply:
I wanted to do that by tagging manga properly then use something like

```    private GetMangaLanguageCode(manga: Manga): string {
        const languageTag: Tag = manga.Tags.Value.find(tag => tag.categoryKey === 'Tags_Language');
        return [...mangaLanguageMap.entries()].find(({ 1: value }) => value === languageTag)[0];
    }
```
But apparently tags are not saved in manga list, meaning 
* Tagging manga is pointless
* i will need to rely on composed Manga id  { id , language}
* And this id will be useless when we finally save media language. Meaning potential breaking with id changes.